LiteCore/Query/SQLiteDataFile+Indexes.cc

```diff
@@ -122,26 +125,68 @@ namespace litecore {
-            stmt.bind(1, tableName);
-            if ( stmt.executeStep() ) return;
+    void SQLiteDataFile::garbageCollectIndexTable(const SQLiteIndexSpec& spec) {
+        string              tableName = spec.indexTableName;  // tableName is in plain form.
+        std::vector<string> unnestTables;
+        if ( spec.type == IndexSpec::kArray ) {
+            for ( size_t pos = 0; pos != string::npos; ) {
+                pos = tableName.find(KeyStore::kUnnestLevelSeparator, pos);
+                if ( pos != string::npos ) {
+                    unnestTables.push_back(tableName.substr(0, pos));
+                    pos += KeyStore::kUnnestLevelSeparator.size;
+                } else {
+                    unnestTables.push_back(tableName);
+                }
+            }
+            // Assertion: unnestTables.back() == tableName
         }
 
-        LogTo(QueryLog, "Dropping unused index table '%s'", tableName.c_str());
-        exec(CONCAT("DROP TABLE " << sqlIdentifier(tableName) << ""));
+        size_t unnestLevel = unnestTables.size();
+        // Invariant: spec.type != IndexSpec::kArray || unnestLevel >= 1
+        while ( true ) {
+            {
+                // Check that if the index table still exists. If so, don't do GC
+                SQLite::Statement stmt(*this, "SELECT name FROM indexes WHERE indexTableName=?");
+                stmt.bind(1, tableName);
+                if ( stmt.executeStep() ) return;
+            }
+            if ( spec.type == IndexSpec::kArray ) {
+                // Assertion: unnestLevel >= 1
+                // check if it has index on child/nested array. If so, we cannot GC
+                SQLite::Statement stmt(*this, "SELECT name FROM indexes WHERE indexTableName like ?");
+                stmt.bind(1, unnestTables[unnestLevel - 1] + string(KeyStore::kUnnestLevelSeparator) + "%");
+                // This table has child table. Don't delete it and its parent
+                if ( stmt.executeStep() ) return;
+
+                tableName = KeyStore::hexName(tableName);  // Now, it's the true table name, hashed.
+                // Move on to delete the unused tables.
+            }
+
+            LogTo(QueryLog, "Dropping unused index table '%s'", tableName.c_str());
+            exec(CONCAT("DROP TABLE " << sqlIdentifier(tableName) << ""));
+
+            static const char* kTriggerSuffixes[]       = {"ins", "del", "upd", "preupdate", "postupdate", nullptr};
+            static const char* kNestedTriggerSuffixes[] = {"ins", "del", nullptr};
+
+            const char** triggerSuffixes = kTriggerSuffixes;
+            if ( unnestLevel > 1 ) triggerSuffixes = kNestedTriggerSuffixes;
+
+            {
+                stringstream sql;
+                for ( int i = 0; triggerSuffixes[i]; ++i ) {
+                    sql << "DROP TRIGGER IF EXISTS \"" << tableName << "::" << triggerSuffixes[i] << "\";";
```

Comment:
This is unsafe if tableName can contain quote characters. Use sqlIdentifier() to quote the trigger name.
(I know this is pre-existing code, but I noticed the issue and it's worth fixing...)

Reply:
okay

Reply:
Cannot do it. Changing it sqlIdentifier(tableName) breaks multiple tests. I think it's already quoted, "EXISTS \"" and "::ins" should be part of the table name.

Reply:
Got it. done